node/src/components/consensus/highway_core/highway.rs

```diff
@@ -574,6 +572,19 @@ impl<C: Context> Highway<C> {
         evidence_effects.extend(self.on_new_unit(&unit_hash, now, rng));
         evidence_effects
     }
+
+    /// Adds endorsements to the state.
+    fn add_endorsements(&mut self, endorsements: Endorsements<C>) -> Vec<Effect<C>> {
+        // TODO: Not faulty — evidence!
+        let old_faulty: HashSet<_> = self.state.faulty_validators().collect();
+        self.state.add_endorsements(endorsements, &self.instance_id);
+        self.state
+            .faulty_validators()
+            .filter(|v_id| !old_faulty.contains(v_id))
```

Comment:
I don't understand what's happening here – isn't `old_faulty` equal to `self.state.faulty_validators()`?

Reply:
OK, now I see it – `state.add_endorsements` may find new evidences which will result in `state.faulty_validators()` having different elements from `old_faulty`.